File: LuoSonglei/week-5/SortColors.c
/*******************************************
Author      : LHearen
E-mail      : LHearen@126.com
Time        : 2016-01-04 08:01
Description : 
Source      : https://leetcode.com/problems/sort-colors/
*******************************************/
#define FOO
void swap(int* p, int* q)
{
    int t = *p;
    *p = *q;
    *q = t;
}

//AC - 0ms;
void sortColors(int* nums, int size)
{
    int p, l, r, q;
    p = l = 0, q = r = size-1;
    while(l <= r)
    {
        while(l < size && nums[l] == 0) l++;
        while(r > -1 && nums[r] == 2) r--;
        if(l < size && nums[l] == 1 && l <= q)
            swap(nums+(l++), nums+(p++)); 
        if(r > -1 && nums[r] == 1 && r >= p)
            swap(nums+(r--), nums+(q--));
        else if(l <= r && nums[r] == 0 && nums[l] == 2)
        {
            swap(nums+r, nums+l);
            r--, l++;
        }
    }
    #ifdef FOO
        for(int i = 0; i < size; i++)
            printf("%d\t", nums[i]);
        printf("\n");
        printf("l: %d\tr: %d\n", l, r);
    #endif
    nums[p] == 1? p : p--;
    nums[q] == 1? q : q++;
    while(l > -1 && p > -1)
        swap(nums+l--, nums+p--);
    while(r < size && q < size)
        swap(nums+r++, nums+q++);
}

void sort(int* nums, int l, int r)
{
    int i = l, j = r;
    int mid = nums[(i+j)>>1];
    while(i <= j)
    {
        mid = (i+j)>>1; 
        while(nums[i] < mid) i++;
        while(nums[j] > mid) j--;
        if(i <= j)
        {
            swap(nums+i, nums+j);
            i++, j--;
        }
    }
    if(l < j)
        sort(nums, l, j);
    if(i < r)
        sort(nums, i, r);
}
//AC - 0ms;
void sortColors1(int* nums, int size)
{
   sort(nums, 0, size-1); 
}

File: CuiGuangfan/src/com/xingkong/SortColors.java
package com.xingkong;

/**
 * @author cuiguangfan 736068048@qq.com:
 * @version create time：2016年1月3日 下午9:22:34 class description
 */
public class SortColors {
	private void swap(int[] nums, int index1, int index2) {
		// 不能使用nums[index1]=nums[index1]^nums[index2];
		// nums[index2]=nums[index1]^nums[index2];
		// nums[index1]=nums[index1]^nums[index2];
		// 当index1=index2时，由于nums[index1]和nums[index2]使用的地址相同，会导致nums[index1]和nums[index2]变为0
		// 如果将值赋值给a,b，由于a和b地址不同，所以不会产生这个问题
		int a = nums[index1];
		int b = nums[index2];
		a = a ^ b;
		b = a ^ b;
		a = a ^ b;
		nums[index1] = a;
		nums[index2] = b;
	}

	public void sortColors(int[] nums) {
		int left = 0;
		int right = nums.length - 1;
		int i = 0;
		while (i <= right) {
			if (nums[i] == 0) {
				swap(nums, left, i);
				i++;
				left++;
			} else if (nums[i] == 2) {
				swap(nums, right, i);
				right--;
			} else {
				i++;
			}
		}
		for (int j = 0; j < nums.length; j++)
			System.out.print(nums[j] + " ");
	}

	private void sortColors_improve(int[] nums) {
		int i = -1;
		int j = -1;
		int k = -1;
		for (int p = 0; p < nums.length; p++) {
			// 根据第i个数字，挪动0~i-1串。
			if (nums[p] == 0) {
				nums[++k] = 2; // 2往后挪
				nums[++j] = 1; // 1往后挪
				nums[++i] = 0; // 0往后挪
			} else if (nums[p] == 1) {
				nums[++k] = 2;
				nums[++j] = 1;
			} else
				nums[++k] = 2;
		}
		for (int kk = 0; kk < nums.length; kk++)
			System.out.print(nums[kk] + " ");
	}

	public static void main(String[] args) {
		int[] nums = new int[] { 1,0,1,1,2,0,1,2 };
		SortColors sc = new SortColors();
//		sc.sortColors(nums);
		sc.sortColors_improve(nums);
	}
}
